x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -732,40 +786,71 @@ private void innerRefresh(String tokenDocId, Authentication userAuth, ActionList
+                                                (Map<String, Object>) supersedingTokenSource.get("refresh_token");
+                                            final String supersedingRefreshTokenValue = (String) supersedingRefreshTokenSrc.get("token");
+                                            reIssueTokens(supersedingUserTokenSource, supersedingRefreshTokenValue, listener);
+                                        } else {
+                                            logger.info("could not find token document [{}] for refresh", supersedingTokenDocId);
+                                            onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                                        }
+                                    }, e -> {
+                                        logger.info("could not find token document [{}] for refresh", supersedingTokenDocId);
+                                        onFailure.accept(invalidGrantException("could not refresh the requested token"));
+                                    }), client::get);
+                            } else {
+                                final Map<String, Object> userTokenSource = (Map<String, Object>)
+                                    ((Map<String, Object>) source.get("access_token")).get("user_token");
+                                final String authString = (String) userTokenSource.get("authentication");
+                                final Integer version = (Integer) userTokenSource.get("version");
+                                final Map<String, Object> metadata = (Map<String, Object>) userTokenSource.get("metadata");
+                                Version authVersion = Version.fromId(version);
+                                try (StreamInput in = StreamInput.wrap(Base64.getDecoder().decode(authString))) {
+                                    in.setVersion(authVersion);
+                                    Authentication authentication = new Authentication(in);
+                                    final String newUserTokenId = UUIDs.randomBase64UUID();
+                                    final Instant refreshTime = clock.instant();
+                                    Map<String, Object> updateMap = new HashMap<>();
+                                    updateMap.put("refreshed", true);
+                                    updateMap.put("refresh_time", refreshTime.toEpochMilli());
+                                    updateMap.put("superseded_by", getTokenDocumentId(newUserTokenId));
+                                    UpdateRequestBuilder updateRequest =
+                                        client.prepareUpdate(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, tokenDocId)
+                                            .setDoc("refresh_token", updateMap)
+                                            .setRefreshPolicy(RefreshPolicy.WAIT_UNTIL);
```

Comment:
I swapped this to immediate; I think it is ok to do that since we expect concurrent attempts to refresh by our top consumer of this api